packages/swingset-liveslots/tools/fakeStorage.js

```diff
@@ -0,0 +1,166 @@
+    has(key) {
+      return map.has(key);
+    },
+    get(key) {
+      return map.get(key);
+    },
+    getNextKey(priorKey) {
+      assert.typeof(priorKey, 'string');
+      ensureSorted();
+      // TODO: binary search for priorKey (maybe missing), then get
+      // the one after that. For now we go simple and slow. But cache
+      // a starting point, because the main use case is a full
+      // iteration. OTOH, the main use case also deletes everything,
+      // which will clobber the cache on each deletion, so it might
+      // not help.
+      const start = priorKeyReturned === priorKey ? priorKeyIndex : 0;
+      let result;
+      for (let i = start; i < sortedKeys.length; i += 1) {
+        const key = sortedKeys[i];
+        if (compareByCodePoints(key, priorKey) > 0) {
+          priorKeyReturned = key;
+          priorKeyIndex = i;
+          result = key;
+          break;
+        }
+      }
+      if (!result) {
+        // reached end without finding the key, so clear our cache
+        clearGetNextKeyCache();
+      }
+      return result;
+    },
+    set(key, value) {
+      if (!map.has(key)) {
+        clearSorted();
+      }
+      map.set(key, value);
+    },
+    delete(key) {
+      if (map.has(key)) {
+        clearSorted();
+      }
+      map.delete(key);
+    },
+  });
+  return fakeStore;
+}
+
+/**
+ * @template [S=Iterable<[string, string]>]
+ * @typedef {object} LiveFakeStorage
+ * @property {KVStore} kvStore
+ * @property {undefined} serialized
+ * @property {() => SerializedFakeStorage<S>} commit
+ * @property {(serialized?: S) => LiveFakeStorage<S>} init
+ */
+
+/**
+ * @template [S=Iterable<[string, string]>]
+ * @typedef {object} SerializedFakeStorage
+ * @property {undefined} kvStore
+ * @property {S | undefined} serialized
+ * @property {() => never} commit
+ * @property {(serialized?: S) => LiveFakeStorage<S>} init
+ */
+
+/**
+ * @template [S=Iterable<[string, string]>]
+ * @typedef {LiveFakeStorage<S> | SerializedFakeStorage<S>} FakeStorage<S>
+ */
+
+export const serializedCommit = harden(() => {
+  throw Error('Cannot commit serialized storage');
+});
+
+/**
+ * @param {Iterable<[string, string]>} [storeMap]
+ * @returns {SerializedFakeStorage}
+ */
+export const initSerializedFakeStorageFromMap = storeMap => {
+  const serialized = new Map(storeMap);
+  return harden({
+    kvStore: undefined,
+    serialized,
+    commit: serializedCommit,
+    // eslint-disable-next-line no-use-before-define
+    init: initLiveFakeStorageFromMap,
+  });
+};
+
+/**
+ * @param {Iterable<[string, string]>} [serialized]
+ * @returns {LiveFakeStorage}
+ */
+export const initLiveFakeStorageFromMap = serialized => {
+  const storeMap = new Map(serialized);
+  const kvStore = makeKVStoreFromMap(storeMap);
+
+  const commit = initSerializedFakeStorageFromMap.bind(null, storeMap);
+
+  return harden({
+    kvStore,
+    serialized: undefined,
+    commit,
+    init: initLiveFakeStorageFromMap,
```

Comment:
They're needed for propagating the kind of storage from one incarnation to another (clone), or the ability to create a new empty storage of the same kind. Open to alternative suggestions but I'd really like to avoid passing around both the storage and the makers.